crates/cli/src/bin/wasm-bindgen-test-runner/headless.rs

```diff
@@ -0,0 +1,444 @@
+use std::cell::{Cell, RefCell};
+use std::env;
+use std::io::{self, Read};
+use std::net::{SocketAddr, TcpListener, TcpStream};
+use std::path::{PathBuf, Path};
+use std::process::{Command, Stdio};
+use std::thread;
+use std::time::{Instant, Duration};
+
+use curl::easy::Easy;
+use failure::{ResultExt, Error};
+use serde::{Serialize, Deserialize};
+use serde_json;
+
+use shell::Shell;
+
+/// Execute a headless browser tests against a server running on `server`
+/// address.
+pub fn run(server: &SocketAddr, shell: &Shell) -> Result<(), Error> {
+    let (driver, args) = Driver::find()?;
+    println!("Running headless tests in {} with `{}`",
+             driver.browser(),
+             driver.path().display());
+
+    // Allow tests to run in parallel (in theory) by finding any open port
+    // available for our driver. We can't bind the port for the driver, but
+    // hopefully the OS gives this invocation unique ports across processes
+    let driver_addr = TcpListener::bind("127.0.0.1:0")?.local_addr()?;
+
+    // Spawn the driver binary, collecting its stdout/stderr in separate
+    // threads. We'll print this output later.
+    shell.status("Spawning Geckodriver...");
+    let mut cmd = Command::new(&driver.path());
+    cmd.args(&args)
+        .arg(format!("--port={}", driver_addr.port().to_string()))
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .stdin(Stdio::null());
+    let mut child = cmd.spawn()
+        .context(format!("failed to spawn {:?} binary", driver.path()))?;
+    let mut stdout = child.stdout.take().unwrap();
+    let mut stderr = child.stderr.take().unwrap();
+    let mut stdout = Some(thread::spawn(move || read(&mut stdout)));
+    let mut stderr = Some(thread::spawn(move || read(&mut stderr)));
+    let print_driver_stdio = Cell::new(true);
+    let _f = OnDrop(|| {
```

Comment:
I feel like all this could be wrapped up in a `BackgroundChild` type like it is/was in the test-project-builder